File: Data Structure & Algorithm/Models/AVL Tree/AVL Tree.ts
class TreeNode<T> {
  public height: number;
  public left: TreeNode<T> | null;
  public right: TreeNode<T> | null;

  constructor(public value: T) {
    this.height = 1;
    this.left = null;
    this.right = null;
  }
}

export class AVLTree<T> {
  private root: TreeNode<T> | null;

  constructor() {
    this.root = null;
  }

  // 插入新节点
  public insert(value: T): void {
    this.root = this.insertAtNode(this.root, value);
  }

  // 在特定节点处插入新值
  private insertAtNode(node: TreeNode<T> | null, value: T): TreeNode<T> {
    if (!node) {
      return new TreeNode(value);
    }

    if (value < node.value) {
      node.left = this.insertAtNode(node.left, value);
    } else if (value > node.value) {
      node.right = this.insertAtNode(node.right, value);
    } else {
      return node; // 重复值不插入
    }

    node.height = 1 + Math.max(this.getHeight(node.left), this.getHeight(node.right));

    // 平衡因子
    let balanceFactor = this.getBalanceFactor(node);

    // 左左情况
    if (balanceFactor > 1 && value < node.left!.value) {
      return this.rightRotate(node);
    }

    // 右右情况
    if (balanceFactor < -1 && value > node.right!.value) {
      return this.leftRotate(node);
    }

    // 左右情况
    if (balanceFactor > 1 && value > node.left!.value) {
      node.left = this.leftRotate(node.left!);
      return this.rightRotate(node);
    }

    // 右左情况
    if (balanceFactor < -1 && value < node.right!.value) {
      node.right = this.rightRotate(node.right!);
      return this.leftRotate(node);
    }

    return node;
  }

  // 右旋转
  private rightRotate(y: TreeNode<T>): TreeNode<T> {
    let x = y.left!;
    let T2 = x.right;

    x.right = y;
    y.left = T2;

    y.height = Math.max(this.getHeight(y.left), this.getHeight(y.right)) + 1;
    x.height = Math.max(this.getHeight(x.left), this.getHeight(x.right)) + 1;

    return x;
  }

  // 左旋转
  private leftRotate(x: TreeNode<T>): TreeNode<T> {
    let y = x.right!;
    let T2 = y.left;

    y.left = x;
    x.right = T2;

    x.height = Math.max(this.getHeight(x.left), this.getHeight(x.right)) + 1;
    y.height = Math.max(this.getHeight(y.left), this.getHeight(y.right)) + 1;

    return y;
  }

  // 获取节点高度
  private getHeight(node: TreeNode<T> | null): number {
    if (!node) {
      return 0;
    }

    return node.height;
  }

  // 计算平衡因子
  private getBalanceFactor(node: TreeNode<T>): number {
    if (!node) {
      return 0;
    }

    return this.getHeight(node.left) - this.getHeight(node.right);
  }

  // 删除节点
  public delete(value: T): void {
    this.root = this.deleteNode(this.root, value);
  }

  private deleteNode(node: TreeNode<T> | null, value: T): TreeNode<T> | null {
    if (node === null) {
      return node;
    }

    if (value < node.value) {
      node.left = this.deleteNode(node.left, value);
    } else if (value > node.value) {
      node.right = this.deleteNode(node.right, value);
    } else {
      // 节点有一个子节点或没有子节点
      if (node.left === null || node.right === null) {
        let temp = node.left ? node.left : node.right;

        if (temp === null) {
          temp = node;
          node = null;
        } else {
          node = temp;
        }
      } else {
        // 有两个子节点的情况
        let temp = this.getMinValueNode(node.right!);
        node.value = temp.value;
        node.right = this.deleteNode(node.right, temp.value);
      }
    }

    if (node === null) {
      return node;
    }

    // 更新高度
    node.height = 1 + Math.max(this.getHeight(node.left), this.getHeight(node.right));

    // 平衡树
    let balanceFactor = this.getBalanceFactor(node);

    // 左左情况
    if (balanceFactor > 1 && this.getBalanceFactor(node.left!) >= 0) {
      return this.rightRotate(node);
    }

    // 左右情况
    if (balanceFactor > 1 && this.getBalanceFactor(node.left!) < 0) {
      node.left = this.leftRotate(node.left!);
      return this.rightRotate(node);
    }

    // 右右情况
    if (balanceFactor < -1 && this.getBalanceFactor(node.right!) <= 0) {
      return this.leftRotate(node);
    }

    // 右左情况
    if (balanceFactor < -1 && this.getBalanceFactor(node.right!) > 0) {
      node.right = this.rightRotate(node.right!);
      return this.leftRotate(node);
    }

    return node;
  }

  // 找到树中最小值的节点
  private getMinValueNode(node: TreeNode<T>): TreeNode<T> {
    let current = node;
    while (current.left !== null) {
      current = current.left;
    }
    return current;
  }

  // 前序遍历
  public preOrderTraversal(): void {
    this.preOrderHelper(this.root);
  }

  private preOrderHelper(node: TreeNode<T> | null): void {
    if (node !== null) {
      console.log(node.value);
      this.preOrderHelper(node.left);
      this.preOrderHelper(node.right);
    }
  }

  // 中序遍历
  public inOrderTraversal(): void {
    this.inOrderHelper(this.root);
  }

  private inOrderHelper(node: TreeNode<T> | null): void {
    if (node !== null) {
      this.inOrderHelper(node.left);
      console.log(node.value);
      this.inOrderHelper(node.right);
    }
  }

  // 后序遍历
  public postOrderTraversal(): void {
    this.postOrderHelper(this.root);
  }

  private postOrderHelper(node: TreeNode<T> | null): void {
    if (node !== null) {
      this.postOrderHelper(node.left);
      this.postOrderHelper(node.right);
      console.log(node.value);
    }
  }
}

File: Data Structure & Algorithm/Archive/Models/AVL Tree/AVL Tree.go
package main

import (
	"fmt"
)

// TreeNode represents a node in the AVL tree
type TreeNode struct {
	value  int
	height int
	left   *TreeNode
	right  *TreeNode
}

// AVLTree represents the AVL tree
type AVLTree struct {
	root *TreeNode
}

// NewAVLTree initializes a new AVL tree
func NewAVLTree() *AVLTree {
	return &AVLTree{}
}

// Insert inserts a new value into the AVL tree
func (tree *AVLTree) Insert(value int) {
	tree.root = tree.insertAtNode(tree.root, value)
}

// insertAtNode inserts a new value at the given node
func (tree *AVLTree) insertAtNode(node *TreeNode, value int) *TreeNode {
	if node == nil {
		return &TreeNode{value: value, height: 1}
	}

	if value < node.value {
		node.left = tree.insertAtNode(node.left, value)
	} else if value > node.value {
		node.right = tree.insertAtNode(node.right, value)
	} else {
		return node // Duplicate values are not inserted
	}

	node.height = 1 + max(tree.getHeight(node.left), tree.getHeight(node.right))

	// Balance the node
	return tree.balanceNode(node, value)
}

// balanceNode balances the given node
func (tree *AVLTree) balanceNode(node *TreeNode, value int) *TreeNode {
	balanceFactor := tree.getBalanceFactor(node)

	// Left Left Case
	if balanceFactor > 1 && value < node.left.value {
		return tree.rightRotate(node)
	}

	// Right Right Case
	if balanceFactor < -1 && value > node.right.value {
		return tree.leftRotate(node)
	}

	// Left Right Case
	if balanceFactor > 1 && value > node.left.value {
		node.left = tree.leftRotate(node.left)
		return tree.rightRotate(node)
	}

	// Right Left Case
	if balanceFactor < -1 && value < node.right.value {
		node.right = tree.rightRotate(node.right)
		return tree.leftRotate(node)
	}

	return node
}

// rightRotate performs a right rotation
func (tree *AVLTree) rightRotate(y *TreeNode) *TreeNode {
	x := y.left
	T2 := x.right

	x.right = y
	y.left = T2

	y.height = max(tree.getHeight(y.left), tree.getHeight(y.right)) + 1
	x.height = max(tree.getHeight(x.left), tree.getHeight(x.right)) + 1

	return x
}

// leftRotate performs a left rotation
func (tree *AVLTree) leftRotate(x *TreeNode) *TreeNode {
	y := x.right
	T2 := y.left

	y.left = x
	x.right = T2

	x.height = max(tree.getHeight(x.left), tree.getHeight(x.right)) + 1
	y.height = max(tree.getHeight(y.left), tree.getHeight(y.right)) + 1

	return y
}

// getHeight returns the height of the given node
func (tree *AVLTree) getHeight(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return node.height
}

// getBalanceFactor returns the balance factor of the given node
func (tree *AVLTree) getBalanceFactor(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return tree.getHeight(node.left) - tree.getHeight(node.right)
}

// Delete deletes a value from the AVL tree
func (tree *AVLTree) Delete(value int) {
	tree.root = tree.deleteNode(tree.root, value)
}

// deleteNode deletes a node with the given value
func (tree *AVLTree) deleteNode(node *TreeNode, value int) *TreeNode {
	if node == nil {
		return node
	}

	if value < node.value {
		node.left = tree.deleteNode(node.left, value)
	} else if value > node.value {
		node.right = tree.deleteNode(node.right, value)
	} else {
		// Node with one child or no child
		if node.left == nil || node.right == nil {
			var temp *TreeNode
			if node.left != nil {
				temp = node.left
			} else {
				temp = node.right
			}

			if temp == nil {
				temp = node
				node = nil
			} else {
				*node = *temp
			}
		} else {
			// Node with two children
			temp := tree.getMinValueNode(node.right)
			node.value = temp.value
			node.right = tree.deleteNode(node.right, temp.value)
		}
	}

	if node == nil {
		return node
	}

	node.height = 1 + max(tree.getHeight(node.left), tree.getHeight(node.right))

	// Balance the node
	return tree.balanceDeleteNode(node)
}

// balanceDeleteNode balances the node after deletion
func (tree *AVLTree) balanceDeleteNode(node *TreeNode) *TreeNode {
	balanceFactor := tree.getBalanceFactor(node)

	// Left Left Case
	if balanceFactor > 1 && tree.getBalanceFactor(node.left) >= 0 {
		return tree.rightRotate(node)
	}

	// Left Right Case
	if balanceFactor > 1 && tree.getBalanceFactor(node.left) < 0 {
		node.left = tree.leftRotate(node.left)
		return tree.rightRotate(node)
	}

	// Right Right Case
	if balanceFactor < -1 && tree.getBalanceFactor(node.right) <= 0 {
		return tree.leftRotate(node)
	}

	// Right Left Case
	if balanceFactor < -1 && tree.getBalanceFactor(node.right) > 0 {
		node.right = tree.rightRotate(node.right)
		return tree.leftRotate(node)
	}

	return node
}

// getMinValueNode finds the node with the minimum value
func (tree *AVLTree) getMinValueNode(node *TreeNode) *TreeNode {
	current := node
	for current.left != nil {
		current = current.left
	}
	return current
}

// PreOrderTraversal performs pre-order traversal of the AVL tree
func (tree *AVLTree) PreOrderTraversal() {
	tree.preOrderHelper(tree.root)
}

func (tree *AVLTree) preOrderHelper(node *TreeNode) {
	if node != nil {
		fmt.Println(node.value)
		tree.preOrderHelper(node.left)
		tree.preOrderHelper(node.right)
	}
}

// InOrderTraversal performs in-order traversal of the AVL tree
func (tree *AVLTree) InOrderTraversal() {
	tree.inOrderHelper(tree.root)
}

func (tree *AVLTree) inOrderHelper(node *TreeNode) {
	if node != nil {
		tree.inOrderHelper(node.left)
		fmt.Println(node.value)
		tree.inOrderHelper(node.right)
	}
}

// PostOrderTraversal performs post-order traversal of the AVL tree
func (tree *AVLTree) PostOrderTraversal() {
	tree.postOrderHelper(tree.root)
}

func (tree *AVLTree) postOrderHelper(node *TreeNode) {
	if node != nil {
		tree.postOrderHelper(node.left)
		tree.postOrderHelper(node.right)
		fmt.Println(node.value)
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	tree := NewAVLTree()
	tree.Insert(10)
	tree.Insert(20)
	tree.Insert(30)
	tree.Insert(40)
	tree.Insert(50)
	tree.Insert(25)

	fmt.Println("Pre-order traversal:")
	tree.PreOrderTraversal()

	fmt.Println("In-order traversal:")
	tree.InOrderTraversal()

	fmt.Println("Post-order traversal:")
	tree.PostOrderTraversal()

	tree.Delete(40)
	fmt.Println("After deleting 40:")
	fmt.Println("In-order traversal:")
	tree.InOrderTraversal()
}